File: apps/blog_agent/db/storage/factory.py
# db/storage/factory.py
"""
Provider factory — reads STORAGE_PROVIDER env and returns a memoized instance.

WHY: Single creation point keeps provider construction side effects isolated
and prevents multiple clients from being instantiated during one process run.
"""

from db.storage.base import StorageProvider

__all__ = ["get_provider"]

_provider: StorageProvider | None = None


def get_provider() -> StorageProvider:
    """Return the process-level StorageProvider singleton.

    Provider is chosen from the ``STORAGE_PROVIDER`` env var:
        - ``supabase`` (default) → SupabaseProvider
        - ``oci`` → OciProvider (boto3 S3-compat, path-style)
        - ``s3``  → OciProvider (same class, different env values)

    Returns:
        A StorageProvider instance.

    Raises:
        ValueError: For unrecognised ``STORAGE_PROVIDER`` values.
        ImportError: If required packages (boto3 / supabase) are missing.
    """
    global _provider
    if _provider is not None:
        return _provider

    from configs.storage import (
        STORAGE_BUCKET,
        STORAGE_PROVIDER,
        SUPABASE_SECRET_KEY,
        SUPABASE_URL,
    )

    provider_name = STORAGE_PROVIDER.lower()

    if provider_name == "supabase":
        from supabase import create_client

        from db.storage.supabase_provider import SupabaseProvider

        client = create_client(SUPABASE_URL, SUPABASE_SECRET_KEY)
        _provider = SupabaseProvider(client=client, bucket=STORAGE_BUCKET)

    elif provider_name in ("oci", "s3"):
        from configs.storage import (
            STORAGE_ACCESS_KEY,
            STORAGE_ENDPOINT,
            STORAGE_REGION,
            STORAGE_SECRET_KEY,
        )
        from db.storage.oci_provider import OciProvider

        _provider = OciProvider(
            endpoint_url=STORAGE_ENDPOINT,
            access_key=STORAGE_ACCESS_KEY,
            secret_key=STORAGE_SECRET_KEY,
            region=STORAGE_REGION,
            bucket=STORAGE_BUCKET,
        )

    else:
        raise ValueError(
            f"Unknown STORAGE_PROVIDER '{STORAGE_PROVIDER}'. "
            "Expected 'supabase', 'oci', or 's3'."
        )

    return _provider


def _reset_provider() -> None:
    """Reset the singleton. Intended for use in tests only."""
    global _provider
    _provider = None

File: apps/dashboard_server/src/lib/storage/factory.ts
// Purpose: resolve the active TStorageProvider from env and memoize it.
//          This is the ONLY module that reads env vars for storage config —
//          all side-effects (env reads) are isolated here.
// Invariants:
//   * getStorageProvider() returns the same instance across calls (memoized).
//   * isStorageConfigured() checks the selected provider has its required env set.
//   * STORAGE_PROVIDER defaults to 'supabase' (backward-compat if env is absent).
// Side effects: reads process.env via the env module (once, at first call).

import {
    STORAGE_PROVIDER,
    SUPABASE_URL,
    SUPABASE_SECRET_KEY,
    SUPABASE_STORAGE_BUCKET,
    STORAGE_BUCKET,
    STORAGE_ENDPOINT,
    STORAGE_ACCESS_KEY,
    STORAGE_SECRET_KEY,
    STORAGE_REGION,
} from '../../config/env.schema';
import type { TStorageProvider } from './provider.type';
import { SupabaseProvider } from './supabaseProvider';
import { OciS3Provider } from './ociS3Provider';

let _provider: TStorageProvider | null = null;

/**
 * Returns the memoized storage provider selected by STORAGE_PROVIDER.
 * Throws if required config for the selected provider is missing.
 */
export function getStorageProvider(): TStorageProvider {
    if (_provider !== null) return _provider;

    const selected = STORAGE_PROVIDER;

    if (selected === 'supabase') {
        if (!SUPABASE_URL || !SUPABASE_SECRET_KEY) {
            throw new Error(
                'supabase provider requires SUPABASE_URL and SUPABASE_SECRET_KEY'
            );
        }
        _provider = new SupabaseProvider({
            supabaseUrl: SUPABASE_URL,
            secretKey: SUPABASE_SECRET_KEY,
            bucket: SUPABASE_STORAGE_BUCKET,
        });
        return _provider;
    }

    // 'oci' and 's3' both use the same S3-compatible client.
    if (selected === 'oci' || selected === 's3') {
        if (
            !STORAGE_ENDPOINT ||
            !STORAGE_ACCESS_KEY ||
            !STORAGE_SECRET_KEY ||
            !STORAGE_REGION
        ) {
            throw new Error(
                `${selected} provider requires STORAGE_ENDPOINT, STORAGE_ACCESS_KEY, STORAGE_SECRET_KEY, STORAGE_REGION`
            );
        }
        _provider = new OciS3Provider({
            endpoint: STORAGE_ENDPOINT,
            region: STORAGE_REGION,
            accessKeyId: STORAGE_ACCESS_KEY,
            secretAccessKey: STORAGE_SECRET_KEY,
            bucket: STORAGE_BUCKET,
        });
        return _provider;
    }

    // TypeScript exhaustiveness guard — should never reach here.
    throw new Error(`unknown STORAGE_PROVIDER: ${String(selected)}`);
}

/**
 * Returns true iff the selected provider has its required env vars set.
 * Used to gate the /images route with a 503 instead of a runtime throw.
 */
export function isStorageConfigured(): boolean {
    const selected = STORAGE_PROVIDER;
    if (selected === 'supabase') {
        return Boolean(SUPABASE_URL && SUPABASE_SECRET_KEY);
    }
    if (selected === 'oci' || selected === 's3') {
        return Boolean(
            STORAGE_ENDPOINT &&
            STORAGE_ACCESS_KEY &&
            STORAGE_SECRET_KEY &&
            STORAGE_REGION
        );
    }
    return false;
}
